# Aufgabe 1

## 1
![image](alphabeta_dark.png)

## 2 
![image](alphabeta_dark_marked.png)

## 3
Der Konoten E und F können vertauscht werden das dazu führt, dass der Knoten E nicht mehr bearbeitet wird

# Aufgabe 2

In [ ]:
import numpy as np

class Spielfeld:

    def __init__(self, field, player, parent):
        self.field = field
        self.player = player
        self.parent = parent

    def expand(self):
        fields : Spielfeld = []
        new_player = 0
        if self.player == 0:
            new_player = 1

        for x in range(3):
            for y in range(3):
                if self.field[x][y] == -1:
                    f = self.field.copy()
                    f[x][y] = self.player
                    fields.append(Spielfeld(f,new_player, self))
        
        return fields

    def value(self):
        f = self.field
        for i in range(3):
            if (f[i, :] == 0).all():
                return 1
            if (f[i, :] == 1).all():
                return -1
            if (f[:, i] == 0).all():
                return 1
            if (f[:, i] == 1).all():
                return -1
        if (f.diagonal() == 0).all():
            return 1
        if (f.diagonal() == 1).all():
            return -1
        if (np.fliplr(f).diagonal() == 0).all():
            return 1
        if (np.fliplr(f).diagonal() == 1).all():
            return -1

        return 0

    def isTerminal(self):
        if self.value() != 0:
            return True

        for x in self.field:
            if -1 in x:
                return False
        
        return True


counter = 0

def Max(field: Spielfeld, alpha, beta):
    global counter
    counter += 1
    if field.isTerminal():
        return field.value(), field

    v = -10000000000
    best = None

    for f in field.expand():
        m, fe = Min(f, alpha, beta)
        if m > v:
            v = m
            best = fe
        if v >= beta:
            return v, best if best is not None else fe

        alpha = max(alpha, v)

    return v, best if best is not None else field



def Min(field: Spielfeld, alpha, beta):
    global counter
    counter += 1
    if field.isTerminal():
        return field.value(), field

    v = 10000000000
    best = None

    for f in field.expand():
        m, fe = Max(f, alpha, beta)
        if m < v:
            v = m
            best = fe
        if v <= alpha:
            return v, best if best is not None else fe

        beta = min(beta, v)

    return v, best if best is not None else field

start = [
    [-1,-1,-1],
    [-1,-1,-1],
    [-1,-1,-1]
]

s = Spielfeld(np.array(start),0, None)

m,fe=Max(s, -10000000, 10000000)

print("Bestes Feld:")
print(fe.field)
print("Wert:", m)
print("Erweiterte Knoten:", counter)


Bestes Feld:
[[0 0 1]
 [1 1 0]
 [0 1 0]]
Wert: 0
Erweiterte Knoten: 18297


Szenario: es wird mit einem leerem spielfeld gestartet
### Ohne alpha beta pruning
549946

### Mit alpha beta pruning
18297

# Aufgabe 3

In [59]:
import numpy as np

counter = 0

def MinMax(field: Spielfeld, alpha, beta, player):
    global counter
    counter += 1

    if field.isTerminal():
        return player * field.value(), field

    v = -100000000
    best = None

    for f in field.expand():
        val, fe = MinMax(f, -beta, -alpha, -player)
        val = -val
        if val > v:
            v, best = val, fe
        alpha = max(alpha, v)
        if alpha >= beta:
            break

    return v, best if best is not None else field



start = np.full((3, 3), -1)
s = Spielfeld(start, 0, None)

m, fe = MinMax(s, -10000000000, 100000000000, +1)

print("Bestes Feld:")
print(fe.field)
print("Wert:", m)
print("Erweiterte Knoten:", counter)


Bestes Feld:
[[0 0 1]
 [1 1 0]
 [0 1 0]]
Wert: 0
Erweiterte Knoten: 18297
